docs/src/main/asciidoc/extension-authors-guide.adoc

```diff
@@ -17,7 +17,53 @@ Quarkus extensions add a new developer focused behavior to the core offering, an
+Without the full and complete context, the best that can be achieved is partial and limited generic support.
+By using the Quarkus extension approach, we can bring high density to Java applications.
+
+The Quarkus extension framework makes a significant impact even when GraalVM is not used (e.g. in HotSpot). 
+Let’s list the actions an extension does:
+
+* An extension hosts Substrate VM code substitution so that libraries can run on GraalVM
+** Most changes are pushed upstream to help the underlying library run on GraalVM
+** Not all changes can be pushed upstream, extensions host Substrate VM substitutions - which is a form of code patching - so that libraries can run
+* Host Substrate VM code substitution to help dead code elimination based on the application needs
+** This is application dependant and cannot really be shared in the library itself
+** For example, Quarkus optimizes the Hibernate code because it knows it only needs a specific connection pool and cache provider
+* Gather build time metadata and generate code
+** This part has nothing to do with GraalVM, it is how Quarkus starts frameworks “at build time”
+** The extension framework facilitates reading metadata, scanning classes as well as generating classes as needed
+* Send metadata to GraalVM for example classes in need of reflection
+** This information is not static per library (e.g. Hibernate) but the framework has the semantic knowledge and knows which classes need to have reflection (for example @Entity classes)
+* Opinionated choices based on the close world view of the application including configuration
+
+=== How to expose configuration
+
+Quarkus simplifies the most common usages.
+This means that its defaults might be different than the library it integrates.
+
+To make the simple experience easiest, unify the configuration in `application.properties` via MicroProfile Config.
+Avoid library specific configuration files, or at least make them optional: e.g. `persistence.xml` for Hibernate ORM is optional.
+
+Extensions should see the configuration holistically as a Quarkus application instead of focusing on the library experience.
+For example `quarkus.database.url` and friends are shared between extensions as defining a database access is a shared task (instead of a `hibernate.*` property for example).
+The most useful configuration options should be exposed as `quarkus.[extension].*` instead of the natural namespace of the library.
+Less common properties can live in the library namespace.
+
+To fully enable the close world assumptions that Quarkus can optimize best, it is better to consider configuration options as build time settled vs overridable at runtime.
+Of course properties like host, post, password should be overridable at runtime.
+But many properties like enable caching, using a given security model (basic, JWT etc) can safely require a rebuild of the application.
+
```

Comment:
So far it's very focused on the native aspect

Do we want to say something about:

* general build time approach
* reactive AND imperative
* where should be the user API (if any)
* where should be the substitutions, templates ...

Also, it might make sense to provide some stereotypes:
* Get a library running - only substitutions
* Get a framework running - move the initialization to the build (Template, configuration, SPI...)
* Migrate a CDI portable extension - or rather how you need to change your extension to work

Reply:
@cescoffier I'm taking this comment into account, I think we should have an imperative + reactive but it's a bit unclear for me how conscise and specific we can be in this paragraph (sebsection really). Can you take a shot whenever?